File: data_structure/SegmentTree/cpp/SegmentTree.cpp
#include "SegmentTree.hpp"

#include <limits>

using namespace std;

namespace {
bool CheckedAdd(long long a, long long b, long long& result) {
    if ((b > 0 && a > numeric_limits<long long>::max() - b) ||
        (b < 0 && a < numeric_limits<long long>::min() - b)) {
        return false;
    }
    result = a + b;
    return true;
}

class SegmentTree {
public:
    explicit SegmentTree(const vector<long long>& initialValues)
        : n_(static_cast<int>(initialValues.size())), tree_(2 * n_, 0) {}

    bool Build(const vector<long long>& initialValues) {
        for (int index = 0; index < n_; ++index) {
            tree_[n_ + index] = initialValues[index];
        }

        for (int node = n_ - 1; node > 0; --node) {
            long long sum = 0;
            if (!CheckedAdd(tree_[2 * node], tree_[2 * node + 1], sum)) {
                return false;
            }
            tree_[node] = sum;
        }

        return true;
    }

    bool Add(int index, long long delta) {
        int node = index + n_;
        long long updatedLeaf = 0;
        if (!CheckedAdd(tree_[node], delta, updatedLeaf)) {
            return false;
        }
        tree_[node] = updatedLeaf;

        for (node /= 2; node > 0; node /= 2) {
            long long sum = 0;
            if (!CheckedAdd(tree_[2 * node], tree_[2 * node + 1], sum)) {
                return false;
            }
            tree_[node] = sum;
        }

        return true;
    }

    bool RangeSum(int left, int right, long long& result) const {
        result = 0;
        int l = left + n_;
        int r = right + n_ + 1;

        while (l < r) {
            if (l & 1) {
                long long updated = 0;
                if (!CheckedAdd(result, tree_[l], updated)) {
                    return false;
                }
                result = updated;
                ++l;
            }
            if (r & 1) {
                --r;
                long long updated = 0;
                if (!CheckedAdd(result, tree_[r], updated)) {
                    return false;
                }
                result = updated;
            }
            l /= 2;
            r /= 2;
        }

        return true;
    }

private:
    int n_;
    vector<long long> tree_;
};
}  // namespace

vector<long long> ProcessSegmentTreeQueries(
    const vector<long long>& initialValues,
    const vector<Query>& queries,
    string& errorMessage
) {
    SegmentTree tree(initialValues);
    if (!tree.Build(initialValues)) {
        errorMessage = "Segment tree overflowed while building the initial state.";
        return {};
    }

    vector<long long> results;
    results.reserve(queries.size());

    for (const Query& query : queries) {
        if (query.type == QueryType::Add) {
            if (!tree.Add(query.left, query.delta)) {
                errorMessage = "Segment tree overflowed while applying an update.";
                return {};
            }
        } else {
            long long sum = 0;
            if (!tree.RangeSum(query.left, query.right, sum)) {
                errorMessage = "Segment tree overflowed while evaluating a sum query.";
                return {};
            }
            results.push_back(sum);
        }
    }

    return results;
}

File: data_structure/SegmentTree/rust/src/segment_tree.rs
#[derive(Clone, Copy)]
pub enum Query {
    Add { index: usize, delta: i64 },
    Sum { left: usize, right: usize },
}

pub struct SegmentTree {
    n: usize,
    tree: Vec<i64>,
}

impl SegmentTree {
    pub fn new(initial_values: &[i64]) -> Result<Self, String> {
        let n = initial_values.len();
        let mut tree = vec![0_i64; 2 * n];
        for (index, value) in initial_values.iter().enumerate() {
            tree[n + index] = *value;
        }
        for node in (1..n).rev() {
            tree[node] = tree[2 * node]
                .checked_add(tree[2 * node + 1])
                .ok_or_else(|| String::from("Segment tree overflowed while building the initial state."))?;
        }
        Ok(Self { n, tree })
    }

    pub fn add(&mut self, index: usize, delta: i64) -> Result<(), String> {
        let mut node = index + self.n;
        self.tree[node] = self.tree[node]
            .checked_add(delta)
            .ok_or_else(|| String::from("Segment tree overflowed while applying an update."))?;

        node /= 2;
        while node > 0 {
            self.tree[node] = self.tree[2 * node]
                .checked_add(self.tree[2 * node + 1])
                .ok_or_else(|| String::from("Segment tree overflowed while applying an update."))?;
            node /= 2;
        }

        Ok(())
    }

    pub fn range_sum(&self, left: usize, right: usize) -> Result<i64, String> {
        let mut result = 0_i64;
        let mut l = left + self.n;
        let mut r = right + self.n + 1;

        while l < r {
            if l & 1 == 1 {
                result = result
                    .checked_add(self.tree[l])
                    .ok_or_else(|| String::from("Segment tree overflowed while evaluating a sum query."))?;
                l += 1;
            }
            if r & 1 == 1 {
                r -= 1;
                result = result
                    .checked_add(self.tree[r])
                    .ok_or_else(|| String::from("Segment tree overflowed while evaluating a sum query."))?;
            }
            l /= 2;
            r /= 2;
        }

        Ok(result)
    }
}

pub fn process_segment_tree_queries(initial_values: &[i64], queries: &[Query]) -> Result<Vec<i64>, String> {
    let mut tree = SegmentTree::new(initial_values)?;
    let mut results = Vec::new();

    for query in queries {
        match *query {
            Query::Add { index, delta } => tree.add(index, delta)?,
            Query::Sum { left, right } => results.push(tree.range_sum(left, right)?),
        }
    }

    Ok(results)
}
